In [1]:
print(type(100))

<class 'int'>


In [2]:
import sys

In [3]:
sys.getsizeof(0)

24

In [4]:
sys.getsizeof(1)

28

In [5]:
sys.getsizeof(100)

28

In [6]:
sys.getsizeof(-1)

28

In [13]:
sys.getsizeof(2**5000)

692

In [14]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [15]:
import time 

In [19]:
def calc(a):
    for i in range(1000000000):
        a * 2

In [20]:
start = time.perf_counter()
calc(6)
end = time.perf_counter()

print(f'Time consimed is {end-start}'.format({end-start}))

Time consimed is 55.49814218200481


# Operations ---> integers support all the standard arithmetic operations .. 

## add + sub - mul * division / (float) power **  

### numerator / demonitor ... 150 / 4 = 38 with remainder 3 ... 155 = 4 * 38 + 3 

### 155 // 4 = 38  and 155 % 4 = 3 >>>> floor division (div // with modular operation mod ... ////) n = d * (n // d + n % d)

### Floor divison is the largest interger less or equal a .. 

# floor(3.14) --> 2
floor(1.999)

In [22]:
from math import floor

In [23]:
floor(1.999)

1

In [24]:
floor(-3.1)

-4

In [25]:
floor(3.14)

3

In [26]:
5 // 3 == floor(5/3)

True

In [31]:
a = -135
b = 4

In [32]:
a = b * (a//b) + a % b
a 

-135

In [33]:
type(1+1)

int

In [34]:
type(4-10)

int

In [35]:
type(3 ** 6)

int

In [36]:
type(2/3)

float

In [37]:
type(10/5)

float

In [38]:
type(10//5)

int

In [39]:
floor(3.5)

3

In [40]:
floor(1.22)

1

In [41]:
floor(-3.00000001)

-4

In [42]:
floor(-3.000000000000000001)

-3

In [43]:
a = 33
b = 16

print(a/b)
print(a//b)
print(floor(a/b))


2.0625
2
2


In [46]:
import math

In [47]:
a = -33
b = 16

print(a/b)
print(a//b)
print(floor(a/b))
print(math.trunc(a/b))

-2.0625
-3
-3
-2


In [48]:

a == b * (a//b) + (a % b)

True

# Representationsn-- encoding schem

In [1]:
encoding = map[digits[0]] + map[digits[1]] + ...
encoding = ''.join([map[d] for d in digits, default]))

NameError: name 'digits' is not defined

In [2]:
int(10.5)

10

In [3]:
int(10.9988
    )

10

In [4]:
int(True), int(False)

(1, 0)

In [5]:
import fractions

In [6]:
a = fractions.Fraction(22, 7)

In [7]:
a

Fraction(22, 7)

In [8]:
print(a)

22/7


In [9]:
float(a)

3.142857142857143

In [10]:
int(a) # turncated 

3

In [11]:
int('123456')

123456

In [12]:
int('101', 2)

5

In [13]:
int('FF', 16)

255

In [15]:
int('ff', 16)

255

In [17]:
int('A', 11)

10

In [18]:
bin(10)

'0b1010'

In [19]:
bin(5)

'0b101'

In [20]:
oct(10)

'0o12'

In [21]:
hex(255)

'0xff'

In [22]:
a = int('101', 2)

In [23]:
a

5

In [24]:
b = 0b101

In [25]:
b

5

In [26]:
def from_base10(n, b):
    if b <2:
        raise ValueError('Base must be >-2')
    if n <0:
        raise ValueError('Number n must be >=0')
    
    if n == 0:
        return [0]
    
    digits = []
    
    while n > 0:
        #m, n = n % b, n//b
        #n = n // b
        n, m = divmod(n, b)
        digits.insert(0, m)
    return digits
        
        
        

In [27]:
from_base10(10, 2)

[1, 0, 1, 0]

In [28]:
from_base10(255, 16)

[15, 15]

In [29]:
def encode(digits, digit_map):
    if max(digits) >= len(digit_map):
        raise ValueError('Digit maps is nog long enough to encode the digits')
    
    encoding = ''
    
    for d in digits:
        encoding += digit_map[d]
    return encoding

In [33]:
encode([11, 4], '0123456789ABCDEF')

'B4'

In [37]:
def encodeJoin(digits, digit_map):
    if max(digits) >= len(digit_map):
        raise ValueError('Digit maps is nog long enough to encode the digits')
    
    return ''.join([digit_map[d] for d in digits])

In [38]:
encodeJoin([11, 4], '0123456789ABCDEF')

'B4'

In [39]:
def rebase_from10(number, base):
    digit_map = '0123456789ABCDEFGHIGKLMNOPQRSTVWXYZ'
    
    if base < 2 or base > 36:
        raise ValueError("Invalied base")
    
    sign = -1 if number < 0 else 1 
    number *= sign
    
    digits = from_base10(number, base)
    enocding = encodeJoin(digits, digit_map)
    
    if sign == -1:
        encoding = '-' + enocding
    return encoding

In [50]:
e = rebase_from10(10, 2)

UnboundLocalError: local variable 'encoding' referenced before assignment

# Rational Number 

A fractions of integer numbers

.45 = 45/100

In [51]:
from fractions import Fraction

In [52]:
x = Fraction(3, 4)
y = Fraction(27, 7)
z = Fraction(6, 10)
x, y ,z

(Fraction(3, 4), Fraction(27, 7), Fraction(3, 5))

In [53]:
print(Fraction(-1, 4))

-1/4


In [55]:
print(Fraction(1,))

1


In [56]:
print(Fraction(2,3) * Fraction(1,2))

1/3


In [57]:
print(Fraction(.75))

3/4


In [58]:
import math 

In [59]:
x = Fraction(math.pi)

In [60]:
x

Fraction(884279719003555, 281474976710656)

In [61]:
print(Fraction(.3))

5404319552844595/18014398509481984


In [4]:
from fractions import Fraction

In [5]:
help(Fraction)

Help on class Fraction in module fractions:

class Fraction(numbers.Rational)
 |  Fraction(numerator=0, denominator=None, *, _normalize=True)
 |  
 |  This class implements rational numbers.
 |  
 |  In the two-argument form of the constructor, Fraction(8, 6) will
 |  produce a rational number equivalent to 4/3. Both arguments must
 |  be Rational. The numerator defaults to 0 and the denominator
 |  defaults to 1 so that Fraction(3) == 3 and Fraction() == 0.
 |  
 |  Fractions can also be constructed from:
 |  
 |    - numeric strings similar to those accepted by the
 |      float constructor (for example, '-2.3' or '1e10')
 |  
 |    - strings of the form '123/456'
 |  
 |    - float and Decimal instances
 |  
 |    - other Rational instances (including integers)
 |  
 |  Method resolution order:
 |      Fraction
 |      numbers.Rational
 |      numbers.Real
 |      numbers.Complex
 |      numbers.Number
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __abs__(a)
 |    

In [6]:
print(Fraction(1))

1


In [7]:
Fraction(1)

Fraction(1, 1)

In [8]:
Fraction(denominator=1, numerator=2)

Fraction(2, 1)

In [9]:
Fraction(numerator=2, denominator=5)

Fraction(2, 5)

In [10]:
Fraction('0.123')

Fraction(123, 1000)

In [11]:
Fraction('2/3')

Fraction(2, 3)

In [12]:
x , y = Fraction(2,3), Fraction(3,4)

In [13]:
x + y

Fraction(17, 12)

In [14]:
Fraction(8, 16)

Fraction(1, 2)

In [15]:
Fraction(1, -4)

Fraction(-1, 4)

In [16]:
x.numerator

2

In [17]:
# irational number
Fraction(2222, 44)

Fraction(101, 2)

In [18]:
import math

In [20]:
x = Fraction(math.pi)

In [21]:
x

Fraction(884279719003555, 281474976710656)

In [24]:
y = Fraction(math.sqrt(2))

In [25]:
y

Fraction(6369051672525773, 4503599627370496)

In [26]:
a = 0.3

In [28]:
print(a)

0.3


In [29]:
Fraction(a)

Fraction(5404319552844595, 18014398509481984)

In [30]:
format(a, '0.5f')

'0.30000'

In [32]:
format(a, '.15f')

'0.300000000000000'

In [33]:
format(a, '.25f')

'0.2999999999999999888977698'

In [34]:
x = Fraction(0.3)

In [35]:
x.limit_denominator(10)

Fraction(3, 10)

In [36]:
x = Fraction(math.pi)

In [37]:
x

Fraction(884279719003555, 281474976710656)

In [38]:
float(x)

3.141592653589793

In [39]:
x.limit_denominator(10)

Fraction(22, 7)

In [40]:
22/7

3.142857142857143

In [41]:
x.limit_denominator(100000)

Fraction(312689, 99532)

In [42]:
312689/99532

3.1415926536189365

# Float ... int can grow .. however float is a fixed number of bytes tha used 8 bytes (64 bits) depneds on C compiler ... how that combile ... using the standard CPython interpreter 

sing > 1 bit 
exponent > 1.5E-5 >>> 1.5 * 1 ^-5 (exponent) 11 bits are in rage [-1022, 1023]
significant digits >> 52 bits >> 15-17 signifiaNT .. 
1.2345 >>> 5 signfance difet .. 12345000000>> 5 significant digirst >>> 12345 * 1 ^-6 

In [43]:
# decimal (base10) representation 19, 3, 5, 6 ... 
for i in range(4):
    ...

In [44]:
1/3

0.3333333333333333

In [45]:
1/3 == 0.3333333333333333

True

# Binary represetnation 
(0.11)_2 = (1/2 + 1/4)_10 
= (1 * 2^-1 + 1 * 2^-2)_10


Similarly,

(0.1101)_2 = (1/2+)...

0.1 = 1/10 

In [46]:
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __ceil__(self, /)
 |      Return the ceiling as an Integral.
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(self, /)
 |      Return the floor as an Integral.
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 

In [47]:
float(10)

10.0

In [48]:
float(10.3)

10.3

In [49]:
float('12.3')

12.3

In [50]:
float('22/3')

ValueError: could not convert string to float: '22/3'

In [51]:
from fractions import Fraction 

In [52]:
a = Fraction('22/4')

In [53]:
a

Fraction(11, 2)

In [54]:
float(a)

5.5

In [55]:
print(0.1)

0.1


In [56]:
format(0.1, '.15f')

'0.100000000000000'

In [57]:
format(0.1, '.44f')

'0.10000000000000000555111512312578270211815834'

In [58]:
1/8

0.125

In [59]:
format(.125, '.25f')

'0.1250000000000000000000000'

In [60]:
a = 0.1 + .1 + .1 

In [65]:
a

0.30000000000000004

In [66]:
b = .3

In [67]:
a == b

False

In [68]:
format(a, '.25f')

'0.3000000000000000444089210'

In [69]:
format(b, '.25f')

'0.2999999999999999888977698'

In [70]:
round(a, 5) == round(b, 5)

True

In [71]:
def is_equal(x, y, eps):
    return math.fabs(x,y) < eps

In [72]:
c = 10000.1 * 3
d = 30000.3

c == d

False

In [73]:
c

30000.300000000003

# The math module has that solution 


In [75]:
math.isclose(a, b)#, *, rel_tol=le-09, abs_tor=0>0)

True

In [76]:
math.isclose(a, b, abs_tol=0)

True

In [77]:
# Float equality testing .. is not neceeary the finit represtenation 

x = 0.1

In [78]:
format(x, '.25f')

'0.1000000000000000055511151'

In [79]:
print(x)

0.1


In [80]:
x = 0.125

In [81]:
format(x, '.25f')

'0.1250000000000000000000000'

In [82]:
x, y

(0.125, Fraction(6369051672525773, 4503599627370496))

In [83]:
round(x, 3) == round(y, 3)

False

In [86]:
x = 10000.01 
y = 10000.02

y / x

1.000000999999

In [87]:
x1 = 0.01
y1 = 0.02

y1 / x1

2.0

In [88]:
round(x, 1)

10000.0

In [89]:
round(y, 1)

10000.0

In [90]:
from math import isclose

In [91]:
help(isclose)

Help on built-in function isclose in module math:

isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
    Determine whether two floating point numbers are close in value.
    
      rel_tol
        maximum difference for being considered "close", relative to the
        magnitude of the input values
      abs_tol
        maximum difference for being considered "close", regardless of the
        magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [92]:
x = 0.1 + 0.1 + 0.1
y = 0.3

In [93]:
isclose(x, y)

True

In [98]:
x = 1234567789.01
y = 1234567789.02

In [100]:
isclose(x, y, rel_tol=0.01)

True

In [101]:
x, y = 0.01, 0.02 
isclose(x,y, rel_tol=0.01)

False

In [102]:
# when dealing with numbers .. when you are very close to zero the relative tolerance will occur a problem ..

isclose(x, y, rel_tol=0.01, abs_tol=0.01)

True

In [103]:
x = 0.0000000001
y = 0.0000000002



In [104]:
a = 123456789.02
b = 123456789.02

In [105]:
print(isclose(x, y, abs_tol=0.0001, rel_tol=0.0001))
print(isclose(a, b, abs_tol=0.0001, rel_tol=0.0001))

True
True


# Truncation -- return integer portion of the number 
# int constructor uses float .. turncation cast ...

# Floor ... is the max of all integer less than or equal to the given number ... a // b == floor (a/b) .. floor()


# ceiling >>> take the smallest integer greater than or equal to the number ... ceil(x)= min{i ,- z | i >=x}

In [107]:
from math import trunc 

In [108]:
help(trunc)

Help on built-in function trunc in module math:

trunc(x, /)
    Truncates the Real x to the nearest Integral toward 0.
    
    Uses the __trunc__ magic method.



In [109]:
trunc(12.444)

12

In [110]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if 

In [111]:
int(13.5)

13

In [112]:
trunc(3.5)

3

In [113]:
from math import floor 

In [114]:
help(floor)

Help on built-in function floor in module math:

floor(x, /)
    Return the floor of x as an Integral.
    
    This is the largest integer <= x.



In [116]:
floor(3.4), floor(3.6), floor(-1.7)

(3, 3, -2)

In [117]:
trunc(3.4), trunc(3.6), trunc(-1.7)

(3, 3, -1)

In [118]:
from math import ceil


In [119]:
help(ceil)

Help on built-in function ceil in module math:

ceil(x, /)
    Return the ceiling of x as an Integral.
    
    This is the smallest integer >= x.



In [120]:
ceil(4.5), ceil(6.4), ceil(-6.4), ceil(-6.7)

(5, 7, -6, -6)

# Round is the closest multiple of 10 ^0 = 1 

# -1 round to the closest multiple of 10 ^-1(-1) = 10

# Rounding away from zero .... round(1.25, 1) = 1.3 and round(-1.25, 1) = -1.3

# Banker's Rounding 

## IEEE 754 standard ... rounds the nearest value with ties rounded to the nearest value with an even least significatn digit 

# bancker rounding is less baise than standard rounding ... as it round to even ..

## int(x + 0.5)

In [121]:
int(4.5 + .5)

5

In [122]:
int(3.4+.5)

3

In [124]:
-  int(abs(4.3 + .5))

-4

In [125]:
int(10.5+.5)

11

In [126]:
int(abs(10.6) + .5)

11

In [127]:
- (int(abs(-10.7) + .5))

-11

In [128]:
math.copysign(1,2)

1.0

In [129]:
def round_up(x):
    from math import fabs, copysign
    return copysign(1, x) * int(fabs(x) + 0.5)

In [132]:
round_up(4.5)

5

In [133]:
round_up(-3.4)

-3

In [131]:
def round_up1(x):
    from math import copysign
    return int(x + copysign(0.5, x))

In [134]:
round_up1(4.5)

5

In [135]:
round_up1(-3.4)

-3

In [136]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



In [137]:
a = round(1.9)
a, type(a)

(2, int)

In [138]:
a = round(1.9, 0)
a, type(a)

(2.0, float)

In [139]:
round(1.888, 3)

1.888

In [140]:
round(1.8888, 2)

1.89

In [141]:
round(1.8888, 0)

2.0

In [142]:
round(1.22228, 4)

1.2223

In [144]:
round(888.88, 1), round(888.88, 0), \
round(888.88, -1), round(888.88, -2), round(888.88, -3), round(888.88, -4)

(888.9, 889.0, 890.0, 900.0, 1000.0, 0.0)

In [145]:
round(9800, -4)

10000

# Ties when rounding to the closest multiple and these are two dis

In [146]:
round(1.35, 1)

1.4

In [147]:
round(1.25, 1)

1.2

In [148]:
round(-1.25, 1)

-1.2

In [149]:
round(-1.35, 1)

-1.4

In [150]:
def _round(x):
    from math import copysign
    return int(x + 0.5 * copysign(1, x))

In [151]:
round(1.5), _round(1.5)

(2, 2)

In [152]:
round(2.5), _round(2.5)

(2, 3)

In [153]:
round(-2.5), _round(-2.5)

(-2, -3)

# Decimal module PEP 327 are 

## Float are infinite 
## Decimals have a context that controls certain aspects of working with decimals 

### Precision ... during arithmetic operations 
### rounding algorithm .. float implement .. our algo ... the closest value .. 

## Gloable context ... and local context .. the default context ... sets temporary settings with out affecting the global settings.. create a local context ..


In [154]:
#

In [155]:
import decimal

help(decimal)

Help on module decimal:

NAME
    decimal - C decimal arithmetic module

MODULE REFERENCE
    https://docs.python.org/3.10/library/decimal.html
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

CLASSES
    builtins.ArithmeticError(builtins.Exception)
        DecimalException
            Clamped
            DivisionByZero(DecimalException, builtins.ZeroDivisionError)
            FloatOperation(DecimalException, builtins.TypeError)
            Inexact
                Overflow(Inexact, Rounded)
                Underflow(Inexact, Rounded, Subnormal)
            InvalidOperation
                ConversionSyntax
                DivisionImpossible
                DivisionUndefined(InvalidOperation, builtins.ZeroDivisionEr

In [158]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [159]:
decimal.localcontext(ctx=None)

In [160]:
# precision and rounding 
ctx = decimal.getcontext()

ctx.prec

28

In [161]:
ctx.rounding

'ROUND_HALF_EVEN'

In [162]:
# global and local context ... 

decimal.getcontext().rounding = decimal.ROUND_HALF_UP

In [163]:

with decimal.localcontext() as ctx:
    ctx.prec = 2
    ctx.rounding = decimal.ROUND_HALF_UP

In [164]:
import decimal



In [165]:
decimal.Decimal(10)

Decimal('10')

In [166]:
from decimal import Decimal

In [167]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [171]:
decimal.getcontext().prec = 6

In [172]:
decimal.getcontext().rounding

'ROUND_HALF_UP'

In [173]:
decimal.getcontext()

Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [174]:
g_ctx = decimal.getcontext()

In [175]:
type(g_ctx)

decimal.Context

In [176]:
g_ctx.rounding = decimal.ROUND_HALF_UP

In [177]:
decimal.getcontext()

Context(prec=6, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [178]:
g_ctx.prec = 28

In [179]:
decimal.Context()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [180]:
decimal.localcontext() # create a new context from a specify or any current decimal context 

In [181]:
type(decimal.localcontext())

decimal.ContextManager

In [185]:
x = Decimal('1.25')
y = Decimal('1.35')

In [186]:
with decimal.localcontext() as ctx:
    print(type(ctx)
          )
    print(ctx)
    print(id(ctx) == id(decimal.getcontext()))
    ctx.prec = 6
    ctx.rounding = decimal.ROUND_05UP
    print(round(x, 1))
    print(round(y, 1))
    

<class 'decimal.Context'>
Context(prec=28, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])
True
1.2
1.3


In [188]:
print(round(x, 1))
print(round(y, 1))

1.3
1.4


# Construction Decimal OBjects 

THe Decimal classs is inthe decimal module 

Decimal(x) 
Decimal('0.1') # stores exactly 0.1 not like floats... 
cant do : Decimal(0.1) >>>> since it will store the float approx ... 0.1000000033

# store string and tuples instead ... 

In [189]:
# using a tuple
#(s, d1,d2,d3, ...), exp) .... s(x) = {0 or 1}

# Context precision and the constructor 

# context precision affects mathematical operations .... but nth e constructor ... 

In [190]:
import decimal

In [191]:
help(Decimal)

Help on class Decimal in module decimal:

class Decimal(builtins.object)
 |  Decimal(value='0', context=None)
 |  
 |  Construct a new Decimal object. 'value' can be an integer, string, tuple,
 |  or another Decimal object. If no value is given, return Decimal('0'). The
 |  context does not affect the conversion and is only passed to determine if
 |  the InvalidOperation trap is active.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __ceil__(...)
 |  
 |  __complex__(...)
 |  
 |  __copy__(...)
 |  
 |  __deepcopy__(...)
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(...)
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(...)
 |      Default object

In [192]:
Decimal(10)

Decimal('10')

In [193]:
Decimal('10.1')

Decimal('10.1')

In [194]:
Decimal(1.0)

Decimal('1')

In [195]:
Decimal('-3.1415')

Decimal('-3.1415')

In [196]:
Decimal((0, (3,1,4,1,5),-4))

Decimal('3.1415')

In [197]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [198]:
Decimal('0.1')

Decimal('0.1')

In [199]:
Decimal(0.1) == Decimal('.1')

False

In [200]:
Decimal(10) == Decimal(10)

True

In [201]:
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_UP, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[Inexact, FloatOperation, Rounded], traps=[InvalidOperation, DivisionByZero, Overflow])

In [205]:
decimal.getcontext().prec = 2

In [209]:
a = Decimal('0.123456789')

b = Decimal('0.123456789')

In [210]:
a, b

(Decimal('0.123456789'), Decimal('0.123456789'))

In [211]:
c = a + b

In [212]:
c

Decimal('0.25')

In [213]:
a = Decimal('0.123456789')

b = Decimal('0.123456789')

print(a, b)

with decimal.localcontext() as ctx:
    ctx.prec = 2
    c = a + b
    print('c within local context: {0}'.format(c))
    

0.123456789 0.123456789
c within local context: 0.25


In [215]:
print('c within local context: {0}'.format(c))

c within local context: 0.25


In [220]:
n = 9
d = 3

In [221]:
n == (n//d) *d + (n%d)

True

10

# orher math function

In [227]:
a = Decimal('0.1')

In [228]:
a.ln()

Decimal('-2.3')

In [229]:
print(a.ln())
print(a.exp())
print(a.sqrt())

-2.3
1.1
0.32


In [230]:
import math

In [231]:
math.sqrt(a)

0.31622776601683794

In [232]:
x = 2
x_dec = Decimal(2)

In [233]:
root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

In [234]:
root_float, root_mixed, root_dec

(1.4142135623730951, 1.4142135623730951, Decimal('1.4'))

In [236]:
root_float*root_float, root_mixed*root_mixed, root_dec*root_dec

(2.0000000000000004, 2.0000000000000004, Decimal('2.0'))

# Memory overhead (by flaot)
# Performance of Decimal is slower than float .. precision .. 


In [237]:
from decimal import Decimal

In [238]:
import sys 

In [239]:
a = 3.1415
b = Decimal('3.1415')

In [240]:
sys.getsizeof(a)

24

In [241]:
sys.getsizeof(b)

104

In [242]:
import time 


def run_flaot(n=1):
    for i in range(n):
        a = 3.1415


def run_decimal(n=1):
    for i in range(n):
        a = Decimal('3.1415')

In [243]:
n = 1000000

In [244]:
start = time.perf_counter()
run_flaot()
end = time.perf_counter()
print('flaot: ', end-start)

flaot:  6.953399861231446e-05


In [245]:
start = time.perf_counter()
run_decimal()
end = time.perf_counter()
print('flaot: ', end-start)

flaot:  0.00017194000247400254


In [246]:
def run_flaot(n=1):
    for i in range(n):
        a + a


def run_decimal(n=1):
    for i in range(n):
        a + a

In [247]:
start = time.perf_counter()
run_flaot()
end = time.perf_counter()
print('flaot: ', end-start)

flaot:  7.228799950098619e-05


In [248]:
start = time.perf_counter()
run_decimal()
end = time.perf_counter()
print('flaot: ', end-start)

flaot:  0.0001532139940536581


In [253]:
def run_flaot(n=1):
    for i in range(n):
        math.sqrt(a)


def run_decimal(n=1):
    for i in range(n):
         math.sqrt(a)

In [254]:
start = time.perf_counter()
run_decimal()
end = time.perf_counter()
print('flaot: ', end-start)

flaot:  6.639100320171565e-05


# Complex number 

In [257]:
help(complex)

Help on class complex in module builtins:

class complex(object)
 |  complex(real=0, imag=0)
 |  
 |  Create a complex number from a real part and an optional imaginary part.
 |  
 |  This is equivalent to (real + imag*1j) where imag defaults to 0.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Convert to a string according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __le__(self, value, /)
 |      Return self<=value.
 |  
 |  __lt__(self, value, /)
 |      Return self<value.

In [258]:
.real, .imag, .conjugate

SyntaxError: invalid syntax (480216032.py, line 1)

In [259]:
d = 2 - 3j

In [261]:
d.real, d.imag, d.conjugate

(2.0, -3.0, <function complex.conjugate()>)

In [262]:
the == and != operators are supported <,>, <=, >= and not 

SyntaxError: invalid syntax (3820804563.py, line 1)

In [263]:
math will not work with complex number ,  instead using cmath , exp, logs, trigs, hyperbolics and inverse hyperbolic, poloar, recgagel,

SyntaxError: invalid syntax (2641241911.py, line 1)

In [264]:
import cmath

In [267]:
cmath.phase(x)

0.0

In [268]:
abs(x)

2

In [270]:
a = -1 + 0j

In [271]:
cmath.phase(a)

3.141592653589793

In [272]:
abs(a)

1.0

In [273]:
cmath.rect(math.sqrt(2), math.pi/4)

(1.0000000000000002+1j)

In [274]:
cmath.isclose(cmath.exp(cmath.pi*1j)+1, 0, abs_tol=0.0001)

True

In [275]:
help(complex)

Help on class complex in module builtins:

class complex(object)
 |  complex(real=0, imag=0)
 |  
 |  Create a complex number from a real part and an optional imaginary part.
 |  
 |  This is equivalent to (real + imag*1j) where imag defaults to 0.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Convert to a string according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash__(self, /)
 |      Return hash(self).
 |  
 |  __le__(self, value, /)
 |      Return self<=value.
 |  
 |  __lt__(self, value, /)
 |      Return self<value.

In [276]:
complex(1, 2)

(1+2j)

In [277]:
a = complex(1, 2)

In [278]:
b = 1 + 2j

In [279]:
a == b

True

In [280]:
a.real, (type(a))

(1.0, complex)

In [282]:
a.conjugate()# THE complete

(1-2j)

In [283]:
a = 1 + 2j

In [284]:
a

(1+2j)

In [285]:
a + b

(2+4j)

In [286]:
a**2

(-3+4j)

In [287]:
a // 2

TypeError: unsupported operand type(s) for //: 'complex' and 'int'

In [288]:
a % 2

TypeError: unsupported operand type(s) for %: 'complex' and 'int'

In [289]:
a = 0.1j

In [290]:
format(a.imag, '.25f')

'0.1000000000000000055511151'

In [292]:
a + a + a == 0.3j

False

In [293]:
format((a+a+a).imag, '.25f')

'0.3000000000000000444089210'

In [294]:
format((0.3).imag, '.25f')

'0.0000000000000000000000000'

In [295]:
True> False

True

In [296]:
1 == True

True

In [297]:
1 is True

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
/var/folders/gk/qmm3q5ld1b74kcdmh5pnm1xw0000gn/T/ipykernel_48639/3482963539.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  1 is True


False

In [298]:
True is 1

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
/var/folders/gk/qmm3q5ld1b74kcdmh5pnm1xw0000gn/T/ipykernel_48639/812578685.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  True is 1


False

In [299]:
True * 100

100

In [300]:
# every class in Python implenet a True ... 

8 == 8

True

In [301]:
bool(0)

False

In [302]:
bool(1)

True

In [303]:
bool(3)

True

In [304]:
bool(-2)

True

In [305]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

In [306]:
issubclass(bool, int)

True

In [307]:
type(True), id(True), int(True)

(bool, 4499322032, 1)

In [308]:
type(False), id(False), int(False)

(bool, 4499322760, 0)

In [309]:
(3<5) == True

True

In [311]:
(3<4) is True

True

In [312]:
1 == 2 == False

False

In [313]:
int(True), int(False)

(1, 0)

In [314]:
1 + True

2

In [315]:
100 - False

100

In [316]:
True > False

True

In [317]:
(True + True + True) % 2

1

In [318]:
True and False

False

In [319]:
True or False

True

In [320]:
bool(0)

False

In [321]:
bool(1)

True

In [322]:
int(True)

1

In [323]:
bool(True)

True

In [324]:
bool(False)

False

In [325]:
bool(-1)

True

In [326]:
bool(0)

False

In [327]:
def __bool__(self):
    return self !=0 

In [328]:
bool(88)

True

In [329]:
bool([1,2,3])

True

In [330]:
bool([])

False

In [331]:
bool(None)

False

In [332]:
if my_list:
    ...

NameError: name 'my_list' is not defined

In [333]:
if mylist_ is not None and len(mylist) > 0:
    ...

NameError: name 'mylist_' is not defined

In [334]:
bool('abc')

True

In [335]:
bool(0)

False

In [336]:
bool(1+0j)

True

In [337]:
bool(0+0j)

False

In [338]:
bool(100)

True

In [339]:
x = 99

In [340]:

bool(x)

True

In [341]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if 

In [342]:
bool(0)

False

In [343]:
bool(0), 0.__bool__()

SyntaxError: invalid decimal literal (1907145867.py, line 1)

In [344]:
help(list)

Help on class list in module builtins:

class list(object)
 |  list(iterable=(), /)
 |  
 |  Built-in mutable sequence.
 |  
 |  If no argument is given, the constructor creates a new empty list.
 |  The argument must be an iterable if specified.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(...)
 |      x.__getitem__(y) <==> x[y]
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self))

In [345]:
a = []

In [346]:
bool(a)

False

In [347]:
a.__len__()

0

In [348]:
bool(a.__len__())

False

In [349]:
b = [2]

In [350]:
bool(b.__len__())

True

In [351]:
bool(0.0), bool(0+0j)

(False, False)

In [352]:
from decimal import Decimal
from fractions import Fraction

In [353]:
bool(Fraction(0, 1)), bool(Decimal('0.0'))

(False, False)

In [355]:
bool(10.5), bool(1j), bool(Fraction(1,2)), bool(Decimal('10.5'))

(True, True, True, True)

In [356]:
a = []
b =''
c = ()

In [357]:
bool(a), bool(b), bool(c)

(False, False, False)

In [358]:
d = {}

In [359]:
s = (1,2)

In [361]:
s1 = set((1,2))

In [362]:
s, d, s1

((1, 2), {}, {1, 2})

In [363]:
bool(None)

False

In [364]:
a = [1, 2, 3]

In [365]:
if a is not None and len(a) > 0:
    print(a[0])
else:
    print('Nothing to see here..')

1


In [367]:
a = None

In [368]:
if bool(a):
    print(a[0])
else:
    print('Nonthing to see here...')

Nonthing to see here...


In [372]:
a = None

In [373]:
if len(a) > 0 and a is not None:
    print(a)
else:
    print('Nothing to see here')

TypeError: object of type 'NoneType' has no len()

In [375]:
if  a is not None and len(a) > 0: # order is matter is not None
    print(a)
else:
    print('Nothing to see here')

Nothing to see here


# Booleans Precedence and Short Circuiting - Lectues

In [376]:
highest to lowest precedence

()
<
>
<=
>=
==
!=
in
is 
not
and
or


SyntaxError: invalid syntax (2406313612.py, line 1)

In [378]:
a , b, x , y = 1, 3, 4, 5

In [379]:
a < b or a > c and not x or y 

True

In [380]:
(a<b) or ((a>c) and (not x) or y)

True

In [381]:
listt = [1,2,3]
s = 2
threshold = 2


In [382]:
if s in listt and price(s) > threshold:
    print('yes')

NameError: name 'price' is not defined

In [383]:
name ... null >> None ... '' 'abc'

if name[0] is string.digits:
    # do something 

SyntaxError: invalid syntax (3913876288.py, line 1)

In [384]:
if name and name[0] in string.digits:
    pass 

NameError: name 'name' is not defined

In [385]:
not --> and --> or 

SyntaxError: invalid syntax (1484705368.py, line 1)

In [386]:
True or True and False

True

In [388]:
True or (True and False)

True

In [389]:
(True or True) and False

False

# Booleans: Precedence and short-circuiting 

Ture or Y ... > True

False and Y --> False

In [390]:
a , b = 10, 2

In [393]:
if a / b > 2:
    print('a {0} is at least twice b {1}'.format(a,b))

a 10 is at least twice b 2


In [400]:
b = None

In [403]:
if b and a/b > 2: # short circuit evaluation 
    print('a is at least twice as b')

In [404]:
if b > 0:
    if a/b >2:
        print('a is at least twice b')

TypeError: '>' not supported between instances of 'NoneType' and 'int'

In [405]:
import string

In [406]:
help(string)

Help on module string:

NAME
    string - A collection of string constants.

MODULE REFERENCE
    https://docs.python.org/3.10/library/string.html
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    Public module variables:
    
    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containi

In [407]:
string.digits

'0123456789'

In [408]:
string.ascii_letters

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [409]:
string.ascii_uppercase

'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [410]:
string.ascii_lowercase

'abcdefghijklmnopqrstuvwxyz'

In [411]:
name = 'Bob'
if name[0] in string.digits:
    print('canot')

In [413]:
name = ''
if name[0] in string.digits:
    print('canot')

IndexError: string index out of range

In [414]:
if len(name)> 0 and name[0] in string.digits:
    print('canot')

In [418]:
name = None
if len(name)> 0 and name[0] in string.digits:
    print('canot')

TypeError: object of type 'NoneType' has no len()

In [419]:
name = None
if name is not None and len(name)> 0 and name[0] in string.digits:
    print('canot')

In [416]:
if len(name) and name[0] in string.digits:
    print('canot')

In [420]:
if name and name[0] in string.digits:
    print('canot')

# Boolean operators 

every object in Python has a truthness value


if x is truth return x, else return y

if x is truth return x, ohterwise evaluates y and retun y

......

## and 
if x is falsy, retun x, otherwise return y


if x is falsy, return x, otherwise evaluates y and return y

## Consequences 

x or y ... if x is truthy return x, otherwise evalutes and return y 

In [421]:
x , y = 1,2

In [422]:
x or y

1

In [423]:
None or x

1

In [424]:
s2=2
s3=4

In [425]:
a = s1 or s2 or s3 or 'N/A'

In [426]:
a

{1, 2}

In [427]:
# Consequences: and .... x and y if x is falsy, returns x, otherwise evaluates and returns y

In [428]:
a = 10 
a = 0


In [429]:
sum , n = 20, 3

In [430]:
av = sum/ n

In [431]:
av

6.666666666666667

In [432]:

if s:
    return s[0]
else:
    return ''

SyntaxError: 'return' outside function (1461322781.py, line 2)

In [433]:
return s and s[0]

SyntaxError: 'return' outside function (2197322469.py, line 1)

In [434]:
return (s and s[0])

SyntaxError: 'return' outside function (1383581065.py, line 1)

In [435]:
The boolean not ... 

not is a built-in function that returns a boolean value 

not x --> True if x 

SyntaxError: invalid syntax (1259454147.py, line 1)

In [436]:
x or y : if x is truth, return  x 

SyntaxError: invalid syntax (1480216876.py, line 1)

In [437]:
'a' or [1, 2]

'a'

In [438]:
'' or [1,2]

[1, 2]

In [439]:
1 or 1/0

1

In [440]:
None or 1/0

ZeroDivisionError: division by zero

In [441]:
0 or 1/2

0.5

In [442]:
s1 = None
s2 = ''
s3 = 'abc'



In [443]:
s1 = s1 or 'n/a'

In [444]:
s2 = s2 or 'n/a'


In [445]:
s3= s3 or 'na'

In [446]:
s1, s2, s3

('n/a', 'n/a', 'abc')

In [447]:
x and y: if x is fulsy: return x, otherwise evaluates and retrun y

SyntaxError: invalid syntax (1702592992.py, line 1)

In [449]:
print(None and 100)

None


In [450]:
[] and [0]

[]

In [452]:
a = 2
b = 0

a/b

ZeroDivisionError: division by zero

In [453]:
a/b in general, return 0 when b is zero

SyntaxError: invalid syntax (1614916383.py, line 1)

In [454]:
if b ==0:
    print(0)
else:
    print(a/b)

0


In [455]:
a = 2
b =0
print(b and a/b)

0


In [456]:
s1 = None
s2 = ''
s3= 'anc'

s3[0]

'a'

In [457]:
s1 and s1[0]

In [458]:
print(s1 and s1[0])

None


In [459]:
print(s1 and s1[0] or '')

In [460]:
print(s1 and s1[0] or '')

In [461]:
print(s2 and s2[0] or '')

In [462]:
print(s3 and s3[0] or '')

a


In [463]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

In [464]:
not True

False

In [465]:
not False

True

In [466]:
bool('as')

True

In [467]:
not bool('abc')

False

In [468]:
not 'abc'

False

In [469]:
not None

True

# Comparison Operators 

identity operations >>>> is is not ---> compares memory address -- any type

value compariosn == != ... compare valules different types ok, but must be compatible 
ordering comparison ... < >= <= > .. does'n work all time
mempership oeprations: in not in


- Note: value equaility operators work between floats and Decimal, but as we have sen 

10.0 == Decimal('10.0')

10.0 == Decimal('10.0')

a == b == c

In [470]:
a == b and b == c

False

In [471]:
a == b  == c

False

In [472]:
1 == Decimal('1.5') == Fraction(3, 2)

False

In [473]:
1 < 2 < 3 

True

In [474]:
1 < 2 and 2 < 3

True

In [475]:
1 < math.pi and math.pi < 3

False

In [476]:
a < b > b

False

In [477]:
a < b and b > b

False

In [478]:
4 < 3 > 10

False

In [479]:
3 < 5 < 6 < 9

True

In [480]:
3<5 and 5<6 and 6<9

True

In [481]:
if my_min == cnt < val > other < myMax not in list:
    ...

NameError: name 'my_min' is not defined

In [482]:
# idenditiy operators 

0.1 is (3+4j)

<>:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
/var/folders/gk/qmm3q5ld1b74kcdmh5pnm1xw0000gn/T/ipykernel_48639/3346986548.py:3: SyntaxWarning: "is" with a literal. Did you mean "=="?
  0.1 is (3+4j)


False

In [483]:
3 is 3

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
/var/folders/gk/qmm3q5ld1b74kcdmh5pnm1xw0000gn/T/ipykernel_48639/30664588.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  3 is 3


True

In [484]:
[1, 2] is [1, 2]

False

In [485]:
[1, 2] == [1, 2]

True

In [486]:
'a' in 'this is a text'

True

In [487]:
3 in [1,2,3]

True

In [488]:
3 not in [1,2,3]

False

In [489]:
'key1' in {'key1': 1}

True

In [491]:
1 in {'key1':1}# check the kdys not athe values

False

In [492]:
3 < 5 

True

In [494]:
1 + 1j < 3 + 4j

TypeError: '<' not supported between instances of 'complex' and 'complex'

In [495]:
from decimal import Decimal
from fractions import Fraction

In [496]:
4 < Decimal('10.5')

True

In [497]:
Fraction(2,3) < Decimal('0.5')

False

In [498]:
4 == 4 +0j

True

In [499]:
True == Fraction(2, 2)

True

In [500]:
True < Fraction(3, 2)

True

In [501]:
1 < 2 < 3

True

In [502]:
1 < 2 and 2 < 3

True

In [503]:
3<1 < 3

False

In [504]:
2<1 < 1/0

False

In [505]:
2<1 and 1<1/0

False

In [506]:
3 < 4 < 1/0

ZeroDivisionError: division by zero

In [507]:
1 < 2 > -5

True

In [508]:
1 < 2 < -5 == Decimal('-5.0')

False

In [510]:
'A' < 'a' <'z' >'Z'

True

In [511]:
'A' < 'a' <'z' >'Z' in string.ascii_letters

True

In [512]:
min = 0
max = 100
age = 44

min < age < max

True